Port build.sh to macOS

Code is in sync with master branch and installs libctest.a to install-deps/libs correctly on Ubuntu systems. But, the sdk fails to build on mac os x. I used os x 10.12.3, but feels the version doesn't matter.  The use of `execute_process()`, which doesn't take into consideration other variables like CMAKE_INSTALL_PREFIX etc is causing problems.

> $:cmake xxxxxxxx$ cmake ..
> -- The C compiler identification is AppleClang 8.1.0.8020041
> -- The CXX compiler identification is AppleClang 8.1.0.8020041
> -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
> -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
> -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- CMake Install Prefix is /usr/local
> -- AIG architecture: x86_64
> -- ctest not found...
> -- Building ctest...
> -- The C compiler identification is AppleClang 8.1.0.8020041
> -- The CXX compiler identification is AppleClang 8.1.0.8020041
> -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
> -- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
> -- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Configuring done
> -- Generating done
> -- Build files have been written to: /Users/xxxxxxxx/src/try/azure-iot-gateway-sdk/deps/ctest/build
> -- Installing ctest. Please wait...
> CMake Error at gatewayFunctions.cmake:170 (message):
>   Error installing ctest: 2
> Call Stack (most recent call first):
>   dependencies-test.cmake:9 (findAndInstall)
>   CMakeLists.txt:201 (include)
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/Users/xxxxxxxx/src/try/azure-iot-gateway-sdk/cmake/CMakeFiles/CMakeOutput.log".
> yyyyyyyy:cmake xxxxxxxx$ 
> yyyyyyyy:cmake xxxxxxxx$ 
> yyyyyyyy:cmake xxxxxxxx$ cat /Users/xxxxxxxx/src/try/azure-iot-gateway-sdk/deps/ctest/build/error.txt 
> CMake Error at cmake_install.cmake:31 (file):
>   file INSTALL cannot copy file
>   "/Users/xxxxxxxx/src/try/azure-iot-gateway-sdk/deps/ctest/build/libctest.a"
>   to "/usr/local/lib/libctest.a".
> 
> 
> make: *** [install] Error 1
> $:cmake xxxxxxxx$ 

The intention of this issue, is not to get production quality sdk binary on mac os x which is not a supported platform for now. The intention is to get the SDK code into XCode/VS editor so that it is easily browseable. A good CMakeLists.txt is needed for cmake-generators([1](https://cmake.org/cmake/help/v3.0/manual/cmake-generators.7.html)) to work.

--- a/jenkins/osx_c.sh
+++ b/jenkins/osx_c.sh
@@ -3,52 +3,16 @@
 # Licensed under the MIT license. See LICENSE file in the project root for full license information.
 #
 
-set -e
+build_root=$(cd "$(dirname "$0")/.." && pwd)
+cd $build_root
 
-script_dir=$(cd "$(dirname "$0")" && pwd)
-build_root=$(cd "${script_dir}/.." && pwd)
-build_folder=$build_root"/build"
-
-CORES=$(sysctl -n hw.ncpu)
-
-# Java binding
-pushd $build_root/bindings/java/gateway-java-binding
-mvn clean install
-popd
-
-# Java proxy gateway
-pushd $build_root/proxy/gateway/java/gateway-remote-module
-mvn clean install
-popd
-
-# Node.js proxy gateway
-pushd $build_root/proxy/gateway/nodejs
-npm install
-npm run lint
-npm test
-popd
-
-# Node.js proxy gateway sample
-pushd $build_root/samples/nodejs_remote_sample
-npm install
-npm run lint
-popd
-
-rm -rf $build_folder
-mkdir -p $build_folder
-pushd $build_folder
-
-cmake \
-    -Ddependency_install_prefix:PATH=$build_root/install-deps \
-    -DOPENSSL_ROOT_DIR:PATH=/usr/local/opt/openssl \
-    -Dbuild_cores=$CORES \
-    -Denable_ble_module:BOOL=OFF \
-    -Denable_java_binding=ON \
-    -Denable_java_remote_modules:BOOL=ON \
-    -Drun_unittests:BOOL=ON \
-    -Drun_e2e_tests:BOOL=ON \
-    ..
-
-cmake --build . -- --jobs=$CORES
-ctest -C "debug" -V
-popd
+OPENSSL_ROOT_DIR=/usr/local/opt/openssl \
+tools/build.sh \
+    --disable-ble-module \
+    --enable-java-binding \
+    --enable-java-remote-modules \
+    --enable-nodejs-remote-modules \
+    --run-unittests \
+    --run-e2e-tests \
+    "$@" #-x
+[ $? -eq 0 ] || exit $?

--- a/tools/build.sh
+++ b/tools/build.sh
@@ -113,25 +113,30 @@ get_cores ()
 {
     CORES=$(grep -c ^processor /proc/cpuinfo 2>/dev/null || sysctl -n hw.ncpu)
 
-    # Make sure there is enough virtual memory on the device to handle more than one job.
-    # We arbitrarily decide that 500 MB per core is what we need in order to run the build
-    # in parallel.
-    MINVSPACE=$(expr 500000 \* $CORES)
+    # The following logic works on Linux, doesn't work on macOS. For now, we'll assume macOS
+    # devices have enough memory per core to run the build, rather than porting this logic.
+    if [[ $(uname -s) != Darwin ]]
+    then
+        # Make sure there is enough virtual memory on the device to handle more than one job.
+        # We arbitrarily decide that 500 MB per core is what we need in order to run the build
+        # in parallel.
+        MINVSPACE=$(expr 500000 \* $CORES)
 
-    # Acquire total memory and total swap space setting them to zero in the event the command fails
-    MEMAR=( $(sed -n -e 's/^MemTotal:[^0-9]*\([0-9][0-9]*\).*/\1/p' -e 's/^SwapTotal:[^0-9]*\([0-9][0-9]*\).*/\1/p' /proc/meminfo) )
-    [ -z "${MEMAR[0]##*[!0-9]*}" ] && MEMAR[0]=0
-    [ -z "${MEMAR[1]##*[!0-9]*}" ] && MEMAR[1]=0
+        # Acquire total memory and total swap space setting them to zero in the event the command fails
+        MEMAR=( $(sed -n -e 's/^MemTotal:[^0-9]*\([0-9][0-9]*\).*/\1/p' -e 's/^SwapTotal:[^0-9]*\([0-9][0-9]*\).*/\1/p' /proc/meminfo) )
+        [ -z "${MEMAR[0]##*[!0-9]*}" ] && MEMAR[0]=0
+        [ -z "${MEMAR[1]##*[!0-9]*}" ] && MEMAR[1]=0
 
-    let VSPACE=${MEMAR[0]}+${MEMAR[1]}
+        let VSPACE=${MEMAR[0]}+${MEMAR[1]}
 
-    if [ "$VSPACE" -lt "$MINVSPACE" ] ; then
-    # We think the number of cores to use is a function of available memory divided by 500 MB
-    CORES2=$(expr ${MEMAR[0]} / 500000) || true
+        if [ "$VSPACE" -lt "$MINVSPACE" ] ; then
+            # We think the number of cores to use is a function of available memory divided by 500 MB
+            CORES2=$(expr ${MEMAR[0]} / 500000) || true
 
-    # Clamp the cores to use to be between 1 and $CORES (inclusive)
-    CORES2=$([ $CORES2 -le 0 ] && echo 1 || echo $CORES2)
-    CORES=$([ $CORES -le $CORES2 ] && echo $CORES || echo $CORES2)
+            # Clamp the cores to use to be between 1 and $CORES (inclusive)
+            CORES2=$([ $CORES2 -le 0 ] && echo 1 || echo $CORES2)
+            CORES=$([ $CORES -le $CORES2 ] && echo $CORES || echo $CORES2)
+        fi
     fi
 }
 

--- a/tools/build.sh
+++ b/tools/build.sh
@@ -19,10 +19,16 @@ enable_native_remote_modules=ON
 enable_nodejs_remote_modules=OFF
 enable_java_remote_modules=OFF
 toolchainfile=
-enable_ble_module=ON
 dependency_install_prefix="-Ddependency_install_prefix=$local_install"
 build_config=Debug
 use_xplat_uuid=OFF
+if [[ $(uname -s) == Darwin ]]
+then
+    # Don't build BLE for macOS, even if the caller doesn't pass `--disable-ble-module`
+    enable_ble_module=OFF
+else
+    enable_ble_module=ON
+fi
 
 usage ()
 {

